examples/otlp/CMakeLists.txt

```diff
@@ -1,43 +1,69 @@
 # Copyright The OpenTelemetry Authors
 # SPDX-License-Identifier: Apache-2.0
 
+if(DEFINED OPENTELEMETRY_BUILD_DLL)
+  add_definitions(-DOPENTELEMETRY_BUILD_IMPORT_DLL)
+endif()
+
 include_directories(
   ${CMAKE_BINARY_DIR}/generated/third_party/opentelemetry-proto
   ${CMAKE_SOURCE_DIR}/exporters/otlp/include)
 
 if(WITH_OTLP_GRPC)
   add_executable(example_otlp_grpc grpc_main.cc)
-  target_link_libraries(
-    example_otlp_grpc ${CMAKE_THREAD_LIBS_INIT} common_foo_library
-    opentelemetry_trace opentelemetry_exporter_otlp_grpc)
+
+  target_link_libraries(example_otlp_grpc ${CMAKE_THREAD_LIBS_INIT}
+                        common_foo_library)
+
+  if(DEFINED OPENTELEMETRY_BUILD_DLL)
+    target_link_libraries(example_otlp_grpc opentelemetry_cpp)
+  else()
+    target_link_libraries(example_otlp_grpc opentelemetry_trace
+                          opentelemetry_exporter_otlp_grpc)
+  endif()
```

Comment:
Two questions here:

- do we really want to put all the sdk and exporter code (which ones, it depends on WITH_JAEGER etc) into the same opentelemetry_cpp library, or should be have one opentelemetry_sdk library and libraries for each exporters ?
- maintaining different libraries dependencies based on platform and compiling options is going to be fragile, and easy to break, causing burden. Any way to improve this ?

Reply:
1. I prefer building SDK/exporter into one binary, so we only need export API and some necessary factory functions from SDK. If building SDK/exporter into separate DLLs, we need export more SDK internal functions for exporter which is fragile and we don't guarantee any stability there.
2. What kind of fragility do you mean? Any concrete example on this? The libraries and dependencies should be in one package and are expected to be updated as a whole.

Reply:
> 1. I prefer building SDK/exporter into one binary, so we only need export API and some necessary factory functions from SDK. If building SDK/exporter into separate DLLs, we need export more SDK internal functions for exporter which is fragile and we don't guarantee any stability there.

Assume two applications.

Application A uses the OTLP GRPC exporter and PROMETHEUS, and the build uses WITH_OTLP_HTTP=NO, WITH_ZIPKIN=NO.

Application B uses the OTLP HTTP exporter and ZIPKIN, and uses WITH_OTLP_GRPC=NO, WITH_PROMETHEUS=NO

How can both applications A and B be deployed on the same machine, when they both use a DLL of the same name, opentelemetry-cpp, and expect different content in it ?

Reply:
>     2. What kind of fragility do you mean? Any concrete example on this? The libraries and dependencies should be in one package and are expected to be updated as a whole.

Assume the example code needs an extra dependency on a example_helpers library.

Because this library is not in opentelemetry-cpp, both target_link_library() calls need to be updated.

Now, the example code also needs opentelemetry_resource.

Because this library is in opentelemetry-cpp, only the second call needs to be updated.

Granted, all incorrect dependencies can and should be caught in CI with a build break, but having twice the amount of target_link_library() in the source code makes it harder to maintain.

Reply:
> > 1. I prefer building SDK/exporter into one binary, so we only need export API and some necessary factory functions from SDK. If building SDK/exporter into separate DLLs, we need export more SDK internal functions for exporter which is fragile and we don't guarantee any stability there.
> 
> Assume two applications.
> 
> Application A uses the OTLP GRPC exporter and PROMETHEUS, and the build uses WITH_OTLP_HTTP=NO, WITH_ZIPKIN=NO.
> 
> Application B uses the OTLP HTTP exporter and ZIPKIN, and uses WITH_OTLP_GRPC=NO, WITH_PROMETHEUS=NO
> 
> How can both applications A and B be deployed on the same machine, when they both use a DLL of the same name, opentelemetry-cpp, and expect different content in it ?

The DLL is considered as common library, so it should container the features which can be used by client. Like here, DLL should include OTLP GRPC/HTTP, ZIPKIN, PROMETHEUS then application A and B can share one DLL. For application with has customized needs like some_unknown_exporter, it is still recommended to use static link instead of DLL here.

Reply:
> > ```
> > 2. What kind of fragility do you mean? Any concrete example on this? The libraries and dependencies should be in one package and are expected to be updated as a whole.
> > ```
> 
> Assume the example code needs an extra dependency on a example_helpers library.
> 
> Because this library is not in opentelemetry-cpp, both target_link_library() calls need to be updated.
> 
> Now, the example code also needs opentelemetry_resource.
> 
> Because this library is in opentelemetry-cpp, only the second call needs to be updated.
> 
> Granted, all incorrect dependencies can and should be caught in CI with a build break, but having twice the amount of target_link_library() in the source code makes it harder to maintain.

For example_helper, it should not add to both target_link_libraries because it is common for both DLL and non-DLL build, so it can be added to target_link_libraries unconditionally.

For the case like opentelemetry-resource, it may need to be added to the second call of target_link_libraries if it is not referenced by any other libraries in the DLL. Currently this is necessary as the DLL dependencies are managed separately. Agree this introduces complexity to maintenance, but assume maintaining the extra dependencies for DLL should be acceptable?

Reply:
> The DLL is considered as common library, so it should container the features which can be used by client. Like here, DLL should include OTLP GRPC/HTTP, ZIPKIN, PROMETHEUS then application A and B can share one DLL. For application with has customized needs like some_unknown_exporter, it is still recommended to use static link instead of DLL here.

To me this is a critical problem, maybe I did not expressed it well enough.

How can 2 vendors ship applications A and B independently ?

Will vendor A see an unwanted exporter in the DLL binary, loaded in application A process, just because vendor B needs it ?

Reply:
DLL linking requires kind of dev package with [import library](https://learn.microsoft.com/en-us/cpp/build/reference/building-an-import-library-and-export-file?view=msvc-170), so A and B links to one DLL, they should be aware that what exporters they could get.

Or if you mean A might load opentelemetry_cpp.dll released with B which is unexpected, app A should either build its own opentelemetry_cpp.dll and release it with A (in the same folder). Any app tries to install opentelemery_cpp.dll to system location should be aware that this may break others and this is a known [DLL hell](https://en.wikipedia.org/wiki/DLL_Hell) issue it it happens.

Reply:
> DLL linking requires kind of dev package with [import library](https://learn.microsoft.com/en-us/cpp/build/reference/building-an-import-library-and-export-file?view=msvc-170), so A and B links to one DLL, they should be aware that what exporters they could get.
> 
> Or if you mean A might load opentelemetry_cpp.dll released with B which is unexpected, app A should either build its own opentelemetry_cpp.dll and release it with A (in the same folder). Any app tries to install opentelemery_cpp.dll to system location should be aware that this may break others and this is a known [DLL hell](https://en.wikipedia.org/wiki/DLL_Hell) issue it it happens.

If each application A and B need to keep a private copy of the opentelemetry-cpp DLL, customized for they own use (choice of exporters), what is the whole point of having DLL then ? Applications A and B could be statically linked instead.

The scenario I am concerned about is DLL stomping, from the link you provided, with a DDL installed in a common location.

Reply:
Install application local DLL is for edge case, and still even a single application could have multiple components(DLL) which can share a single copy of the DLL built by the app.

Install a DLL to a common location should be well considered because it will then be considered as kind of system components and should fulfil most user requirements. Like if Windows ship opentelemetry_cpp.dll into c:\windows\system32\, then the users can link their app to it (if the required exporter is included there then linking will succeed). If the link failed because missing exporters, the user has to built their own DLL or use static linking.